src/coreclr/hosts/corerun/corerun.cpp

```diff
@@ -551,308 +329,297 @@ HRESULT InitializeHost(
-    hr = ExecuteAssembly(log, hostEnvironment, managedAssemblyFullName, hostHandle, domainId, argc - 1, (argc - 1) ? &(argv[1]) : NULL, (unsigned int*)&exitCode);
-    if (FAILED(hr))
-    {
-        return false;
-    }
+    return false;
+}
 
-    log << W("App exit value = ") << exitCode << Logger::endl;
+static int self_test();
 
-    hr = ShutdownHost(log, hostEnvironment, hostHandle, domainId, (unsigned int*)&exitCode);
-    if (FAILED(hr))
-    {
-        return false;
-    }
+//
+// Entry points
+//
 
-    return true;
-}
+int MAIN(const int argc, const char_t* argv[])
+{
+    configuration config{};
+    if (!parse_args(argc, argv, config))
+        return EXIT_FAILURE;
+
+    if (config.self_test)
+        return self_test();
 
-void showHelp() {
-    ::wprintf(
-        W("Runs executables on CoreCLR\r\n")
-        W("\r\n")
-        W("USAGE: coreRun [/d] [/v] Managed.exe\r\n")
-        W("\r\n")
-        W("  where Managed.exe is a managed executable built for CoreCLR\r\n")
-        W("        /v causes verbose output to be written to the console\r\n")
-        W("        /d causes coreRun to wait for a debugger to attach before\r\n")
-        W("         launching Managed.exe\r\n")
-        W("\r\n")
-        W("  CoreCLR is searched for in %%core_root%%, then in the directory\r\n")
-        W("  that coreRun.exe is in, then finally in %s.\r\n"),
-        coreCLRInstallDirectory
-        );
+    int exit_code = run(config);
+    return exit_code;
 }
 
-int __cdecl wmain(const int argc, const wchar_t* argv[])
+#ifdef TARGET_WINDOWS
+// Used by CoreShim to determine running CoreCLR details.
+extern "C" __declspec(dllexport) HRESULT __cdecl GetCurrentClrDetails(void** clrInstance, unsigned int* appDomainId)
 {
-    // Parse the options from the command line
-
-    bool verbose = false;
-    bool waitForDebugger = false;
-    bool helpRequested = false;
-    int newArgc = argc - 1;
-    const wchar_t **newArgv = argv + 1;
+    assert(clrInstance != nullptr && appDomainId != nullptr);
+    *clrInstance = CurrentClrInstance;
+    *appDomainId = CurrentAppDomainId;
+    return S_OK;
+}
+#endif // TARGET_WINDOWS
 
-    auto stringsEqual = [](const wchar_t * const a, const wchar_t * const b) -> bool {
-        return ::_wcsicmp(a, b) == 0;
-    };
+//
+// Tests
+//
 
-    auto tryParseOption = [&](const wchar_t* arg) -> bool {
-        if ( stringsEqual(arg, W("/v")) || stringsEqual(arg, W("-v")) ) {
-                verbose = true;
-                return true;
-        } else if ( stringsEqual(arg, W("/d")) || stringsEqual(arg, W("-d")) ) {
-                waitForDebugger = true;
-                return true;
-        } else if ( stringsEqual(arg, W("/?")) || stringsEqual(arg, W("-?")) || stringsEqual(arg, W("-h")) || stringsEqual(arg, W("--help")) ) {
-            helpRequested = true;
-            return true;
-        } else {
-            return false;
+#define THROW_IF_FALSE(stmt) if (!(stmt)) throw W(#stmt);
+static int self_test()
```

Comment:
Could you add a comment what is the purpose of this? It's not clear from the code (at least to me) and you didn't document it in the usage.

Reply:
This is to have a place for some testing if needed. I can remove it if that doesn't make sense. Opinions welcome. If it does make sense I will comment.

Reply:
I don't mind it - I just feel the "self-test" is not good enough description 😉